hpx/parallel/algorithms/for_each.hpp

```diff
@@ -45,11 +46,33 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
             typename hpx::util::decay<F>::type& f_;
             typename hpx::util::decay<Proj>::type& proj_;
 
+            template <typename T>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            typename std::enable_if<
+               !hpx::traits::is_value_proxy<T>::value
+            >::type
+            call(T && t)
+            {
+                hpx::util::invoke(f_,
+                    hpx::util::invoke(proj_, std::forward<T>(t)));
+            }
+
+            template <typename T>
+            HPX_HOST_DEVICE HPX_FORCEINLINE
+            typename std::enable_if<
+                hpx::traits::is_value_proxy<T>::value
+            >::type
+            call(T && t)
+            {
+                auto tmp = hpx::util::invoke(proj_, std::forward<T>(t));
```

Comment:
After several compilation tests, in [https://github.com/STEllAR-GROUP/hpx/blob/master/examples/compute/cuda/partitioned_vector.cu](url) at line 32, Cuda Clang (probably in the device compilation step) seem to deduce the template argument `T` as an `int` and we want it to be a `hpx::compute::cuda::value_proxy<int>`. At the last forwarding operation (a.k.a passing the call operator of invoke_impl in [https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/util/invoke.hpp)](url) line 33), I checked that the template parameter pack `Ts...` is effectively `hpx::compute::cuda::value_proxy<int> &`. @hkaiser, any ideas?